sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/PostgresConnectionProvider.scala

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.jdbc.connection
+
+import java.sql.{Connection, Driver}
+import java.util.Properties
+import javax.security.auth.login.{AppConfigurationEntry, Configuration}
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions
+import org.apache.spark.sql.execution.datasources.jdbc.connection.PostgresConnectionProvider.PGJDBCConfiguration
+import org.apache.spark.util.SecurityUtils
+
+private[jdbc] class PostgresConnectionProvider(driver: Driver, options: JDBCOptions)
+    extends BasicConnectionProvider(driver, options) {
+  def setAuthenticationConfigIfNeeded(): Unit = {
+    val parent = Configuration.getConfiguration
+    val configEntry = parent.getAppConfigurationEntry(appEntry)
+    if (configEntry == null || configEntry.isEmpty) {
+      val config = new PGJDBCConfiguration(parent, appEntry, options.keytab, options.principal)
+      Configuration.setConfiguration(config)
+    }
+  }
+
+  override def getConnection(): Connection = {
+    setAuthenticationConfigIfNeeded()
+    super.getConnection()
+  }
+
+  def appEntry: String = {
+    val parseURL = driver.getClass.getMethod("parseURL", classOf[String], classOf[Properties])
+    val properties = parseURL.invoke(driver, options.url, null).asInstanceOf[Properties]
```

Comment:
This feels a bit yucky but I couldn't find an easily available method to parse the query string... (you could use `URI.getQuery().split("&").find(...)` though.)

Reply:
I've tried this before but not parsing it properly so abandoned:
```
scala> new java.net.URI("jdbc:postgresql://localhost/postgres?jaasApplicationName=custompgjdbc").getQuery()
res1: String = null
```
I was able to solve this with more code where I've concluded that custom implementation doesn't worth.

Reply:
> This feels a bit yucky

I agree but don't have less yucky idea...
`...split("?")...split("&").find(...)` would be one possibility but I think it would be brittle in a different way...

Reply:
Ok... even though using reflection to access private methods in external libraries is always a brittle solution to anything, if that becomes a problem this can be rewritten easily.